packages/voting-contract/src/lib.rs

```diff
@@ -116,12 +116,14 @@ where
 {
     // ensure proposal exists and can be voted on
     let mut prop = proposals().load(deps.storage, proposal_id)?;
-    if prop.status != Status::Open {
+    prop.update_status(&env.block);
+    proposals::<P>().save(deps.storage, proposal_id, &prop)?;
```

Comment:
So if I understand this correctly, you save here so that the status is updated even if an error is returned from the fn in the next lines?

Here's the thing. This save doesn't write to the storage directly, but to a cache. All such changes are committed to persistent storage later if the fn succeeds and discarded if the fn returns an error. So basically - this save will achieve nothing when there's an error returned.

I suggest just having one proposal save in the entire function.

Reply:
Interesting.. 
https://github.com/CosmWasm/cosmwasm/blob/main/packages/std/src/traits.rs#L22
`Storage` description itself doesn't tell you anything about the cache.
I didn't know about it.